Event display updated 2

WenzDaniel opened the pull request:
**What is the problem / what does the code in this PR do**
In this updated we add some flexibility to the interactive event display. The following **optional** features (and their combinations) are now supported:

* Only main/alt S1/S2 in waveform plot
* Only main S1/S2 in peak detail plots, PMT array and waveform plot
* Removing bottom PMT array plot.  

After the feedback we also changed:

* Units of detailed peak plots
* PMTs array color
* Detailed peak appearance
* dt field in mousover

 Further, we added the possibility to display an interactive record matrix below the waveform plot. Last but not least we add an interactive data selector which makes selecting populations easier and quicker. This feature does not replace proper cuts, but helps to quickly study event populations during commissioning. 

All new features are explained in the corresponding [analysis notebook](https://github.com/XENONnT/analysiscode/blob/master/StraxTests/EventDisplayInteractive.ipynb).

jorana commented:
Hi Daniel, can I already ask for a few features to be added:
- straxen.activate_bokeh / straxen.activate_holovies / straxen.activate_interactive (do the magic commands to run as a bokeh server, I always first need to find your examples, github doesn't render, etc.)
- straxen.interactive_to_wiki (an easy function to convert the HTML file to the wiki format). Perhaps even dump html to a textfile for multiple waveforms.
- straxen.save_interactive_as

WenzDaniel commented:
@jorana I implemented your first two suggestions. The third one is not trivial as you need different packages depending on the browser the user is using [(see also here)](https://docs.bokeh.org/en/latest/docs/user_guide/export.html).  But I added this information to the example notebook.

WenzDaniel commented:
@kdund as I understood from your private message you also made some similar changes. Since I do not want to interfere I would suggest you commit your changes and then I add all the things which are still missing (which are none I think)?

kdund commented:
Hi, @WenzDaniel , sorry I did not see the mention on github before now-- only my workplace mail was notifying, so lost this notification. 
Did pull it and it looks good though!

skazama commented:
Hi Daniel, thanks for the update! Here are my comments

- Only main/alt S1/S2 in waveform plot: this is fine!
- Only main S1/S2 in peak detail plots, PMT array and waveform plot: I think still alt_s1/s2 information is quite useful for example like [this](https://xenonnt.slack.com/archives/C016USUN64A/p1615587601078400) or for understanding double-scatter properties, why are we removing these information now?
- Removing bottom PMT array plot: this is also quite important information (I think @feigaodm proposed to add it)

For official plot, maybe we don't want to show redundant information, but for ourselves such information is still useful to understand detector.

WenzDaniel commented:
@skazama ah sorry, I wrote this a bit misleading. All the "changes" I mentioned are in reality options. So you can deactivate/activate certain features as you like. E.g. If you would like to see only the main/alt S1/S2 in the waveform plot you can simply put   `only_peak_detail_in_wf=True` if you set in addition `only_main_peaks=True` you will have only the main S1/S2 in the waveform plot as well as peak detail plots. A detailed notebook can be found [here](https://github.com/XENONnT/analysiscode/blob/master/StraxTests/EventDisplayInteractive.ipynb).

skazama commented:
Thanks Daniel. Everything is clear now to me

kdund commented:
For me the notebook works good (we might want a log-y-scale? I can propose code tmr) 
I could run the example notebook except the 1T result

WenzDaniel commented:
You mean for all three, the S1/S2 detail and waveform plot? I can add those changes, should be just something like 
```
fig.yaxis.y_axis_type="log"
```
or equivalent.

jmosbacher reviewed (approved):
This is a very nice little tool. In general I think since this is not core code but rather a utility function there is no reason not to merge the code and make incremental improvements while its already in use. 

General comments
--------------------------

-  I think using holoviews is a very good idea, it helps abstract and simplify the code in way that increases maintainability since its very expressive and intuitive. I would try to use it almost exclusively and minimize using bokeh directly as much as possible for that reason. 
- Consider using param.Parameterized classes with panel to increase maintainability and organization of the code and logic.
- holoviews streams coupled with DynamicMap are very powerful tools for this kind plotting, consider allowing event selection and generating the waveform plot from the selected event.

Specific
-----------

- I would not use the bokeh Div at all, it does not play well with plots (see attached screenshots) i would probably recommend using holoviews title attribute or a panel markdown for the title text. Wrapping the div in a bokeh widgetbox may also solve the layout problem.
Placing the bokeh column inside a panel pane:
![straxen_interactive_waveforms](https://user-images.githubusercontent.com/21005561/112035460-40713800-8b48-11eb-832e-c6a4b124940b.png)
After removing the div:
![straxen_interactive_waveforms_no_div](https://user-images.githubusercontent.com/21005561/112035458-3f400b00-8b48-11eb-8df7-eaa7a46c01d3.png)

WenzDaniel commented:
Hi Yossi, thanks for your review. About your general comments: Yes I tend to agree, however I learned first about bokeh before I had a closer look into holoviews so there was some "historical" development. About param and panel I do not know much about them, but if I have a bit time in the future I will look into it. 
About you more specific comment: That is interesting I checked this on 3 different machines with 3 different browsers, but never experience this issue. But as the holoviews record matrix extension also requires the usage of panel I will change it also for the standard display before merging.

WenzDaniel commented:
Hi Yossi, are you sure that the issue is related to the DIV object in general? I tried to reproduce your error put I was not successful (neither with firefox nor with chrome). Further, I checked your proposed solution with panel columns, but this solution uses under the hood bokeh DIV too. E.g. if I use 
```
event_display = panel.Column(title, plot,
                                 sizing_mode='scale_width')
```
in which title is just a simple markdown string I get 
```
event_display .objects 
> [Bokeh(Div), Bokeh(Column)]
```

jmosbacher commented:
@WenzDaniel Sorry I should have specified that this issue came up in a jupyter notebook, so most likely its a layout overflow issue where the size of the container is not set to be large enough to hold all the contents (being contracted by the size of the jupyter cell), so they overlap. I did not investigate this too much, just noticed that removing the div removed the issue. Of course in the end everything is converted to bokeh objects but in my experience panel does a better job at laying things out without these weird stuff happening so i try to use it instead of creating the bokeh objects myself. But like I said, this is not very important and I see no reason not to merge.
